[asl][reference] Remove inlining of clist and storage_keyword

IGGeorgiev opened the pull request:
This patch removes all inline derivations except for `binop` and `unop`, which need to be inlined for the lr(1) precedence derivations.

Some new production rules were added to resole shift/reduce conflicts explicitly.

IGGeorgiev commented on asllib/Parser.mly:
```diff
@@ -450,11 +454,13 @@ let local_decl_keyword :=
   *)
 
-let storage_keyword ==
+let storage_keyword :=
```
I mirrored `local_decl_keyword` which does the same thing.  I got the comment from it as well.

I've updated the comment to clarify which productions were conflicting for both this case and the E_Record cases.
I haven't renamed this production yet. I'm not actually sure what a good name for it would be and whatever we change it to might want to be mirrored in `local_decl_keyword`. We could add `non_var_..` as a prefix? That feels a bit noisy to me.

IGGeorgiev replied:
I've appended `_non_var` to the end of them and updated their corresponding macro renderings in the tex as well as their `build_<name>` functions for ast construction for consistency.

I haven't renamed the macro rules themselves and I haven't touched any adjacent `ASTRule` titles (e.g. `ASTRule.SotrageKeyword`). Should I update those as well or is updating the production name sufficient?

Roman-Manevich reviewed (commented):
I don't see the updates to the reference for `expr_pattern`.

IGGeorgiev commented on asllib/Parser.mly:
```diff
@@ -443,18 +449,20 @@ let untyped_decl_item :=
   *)
 
-let storage_keyword ==
+let storage_keyword_non_var :=
```
I've done this, I've also taken the liberty of doing a bit of naming alignment and ran a few `sed` scripts:
 * renamed `\Nstoragekeyword` to `\Nglobaldeclkeyword`
 * renamed `\buildstoragekeyword` to `\buildglobaldeclkeyword`
 * updated the link definitions and uses for the above to use globaldeclkeyword instead of storagekeyword
 * updated ASTRule.StorageKeyword to ASTRule.GlobalDeclKeyword everywhere.
 
 I've also added `expr_pattern`'s empty record case (my bad I didn't realize patterns have their own file, I was wondering where expr_pattern was, I assumed there was prose signifying that `expr_pattern` builds the same as `expr` in all cases)

Roman-Manevich reviewed (approved):
Looks good to me!

IGGeorgiev commented:
Note: I added one extra commit to inline the `end_semicolon` production, I think I removed the inlining of that by accident. It's only there to make it easier to annotate the `allow_no_semicolon` case, otherwise it wouldn't be separated at all. This keeps it in line with the spec, which inlines it.

Roman-Manevich commented on asllib/Parser.mly:
```diff
@@ -184,7 +184,7 @@ let nlist(x) :=
   | ~=x; l=nlist(x); { x :: l }
 
-let end_semicolon :=
+let end_semicolon ==
```
I wonder why we're not mirroring this non-terminal in the documentation.

IGGeorgiev replied:
I added it explicitly to handle the backwards compatibility issue of an optional semicolon easier.

I didn't think it necessary to document it as it should really just be a `terminated_by(END; SEMI_COLON; ..)` once that flag is no longer relevant.
